text/0000-katello-as-a-plugin.md

```diff
@@ -0,0 +1,107 @@
+
+Another aspect of certificates is the desire by users to be able to supply their own certificates for the web UI. Users should be able to change those fluidly without having to propagate new certificates across their entire client infrastructure.
+
+#### Design
+
+Introduce the concept of an infrastructure port that the application runs on and uses the self-signed certificates for. This port would exist in a known range (e.g. 9070 - 9079) so that all other backend services could be similarly configured within this range with well defined port values. For example, Pulp could be deployed on 9072 and Candlepin 9073. All Smart Proxies, Capsules and Clients would talk to port 9070 and leave the web UI to run on 443 as is today. This would allow users to switch out the web UI certificates simply and easily while requiring no updates to the clients. Some discussion of this can be seen [here](https://github.com/theforeman/puppet-foreman/pull/374).
+
+#### Drawbacks
+
+This could introduce potential performance issues with running the application on two ports or complex configuration to ensure proper redirects. This change would introduce a one-time infrastructure update for existing users to move servers and Smart Proxies to the new port and certificate setup. Clients would also need updating to ensure they are pointing and talking to the correct port with RHSM. 
+
+## Smart Proxy vs. Capsule
+
+A Capsule is essentially a Smart Proxy that deploys Pulp, and a reverse proxy to achieve content federation and an isolated communication point fo clients. Foreman deploys the Smart Proxy on port 8443 by default, while Katello deploys it on 9090 by default. Katello deployed Smart Proxies on 9090 because Candlepin was already running on 8443 on the server.
+
+#### Design
+
+Given the above discussion of infrastructure port, introduce a designated Smart Proxy port (e.g. 9071) that is predictable and not commonly used as 8443 is today. Further, merge the puppet-capsule and puppet-foreman_proxy modules to unify Smart Proxies and Capsules and give the content aspects of a Smart Proxy a new name (e.g. content node or content proxy).
+
+#### Drawbacks
+
+The port change would require infrastructure propogation for existing users. The current Capsule concept does not use a Smart Proxy in the traditional sense given it uses a reverse proxy for clients to talk back to the server in an isolated way and deploys Pulp for content federation. Given Pulp has it's own stable API, there is no need to go through the Smart Proxy. However, the Smart Proxy is used to provide statusof Pulp, storage information and indicators of enabled features.
+
```

Comment:
I think we need to think of the smart proxy as two different parts:
1. the smart proxy service it self  (with its api's)
2. the way foreman looks at smart proxies (location, organizations, features etc).

It make sense that every backend service is a smart proxy (from option 2), and probably not every backend needs to be transformed into a smart proxy plugin (1) - yet, in some cases it make sense, it could also be used as a way to act on the backend without a service (think pulp migrations).